### DISCUSSION TITLE
How to deploy Nightscout LibreLinkUp Uploader on Google Cloud?

### DISCUSSION THREAD

**Discussion by mduiella** [NightscoutFoundation/xDrip#3067] [Q&A]
Hello,
my wife has been using FreeStyle 2 sensors with XDrip4iOS and Nightscout on Google Cloud (having moved from Railway and previously Heroku).
Now, she is being "upgraded" to FreeStyle 3 (we live in Belgium). I have been searching for ways to continue using xdrip with FS3, and found the workaround of having Nightscout receive the data from LibreLinkUp, and setting XDrip4iOS as follower of Nightscout. My only problem is that, with almost zero programming knowledge, I don't know how to add the Nightscout LibreLinkUp Uploader (https://github.com/timoschlueter/nightscout-librelink-up) to my wife's Nightscout which runs on Google Cloud (assuming that this is even possible...). 
So: is this technically feasible? If yes, how can it be done?

If it's not feasible/too complex/messy, I am also happy to just go back to Heroku, and pay them what they deserve.

Many thanks in advance,
Matteo

PS: My wife activated her last FS2 sensor last night, so I have 14 days to fix this :)
PPS: I am posting this message here also because I imagine that many T1D persons (as well as their parents, partners, etc.) around the EU will be in a similar situation as FS3 progressively replaces FS2.

**Comment by Navid200** [comment]
The Nightscout that you run in Google Cloud Nightscout, by default, is the same Nightscout you would run on Heroku.
If you want, you can customize your Nightscout.  But, by default, it is identical. 

If anyone has ever told you that you cannot do something with Google Cloud Nightscout that you could with Heroku, or any other Nightsocut server, they have intentionally or unintentionally misled you.

If you know how to enable an uploader on Heroku and doing the same fails on Google Cloud Nightscout, I would love to know the details.

**Reply by mduiella** [reply]
Dear @Navid200 
thanks for the quick reply!
Yes, customising my Nightscout (by adding the LibreLinkUp uploader) is precisely what I would like to do.
Unfortunately, my problem is that I lack basic IT skills... hence, I do not know how to enable an uploader. I set up the Nightscout on Google Cloud by following your excellent step-by-step guide, but I cannot do anything else beyond that... 
Simply, on the repository page of the LLU uploader, I see that there is a nice "Deploy to Heroku" button - which (I suppose) could allow someone as incompetent as me to deploy it on Heroku sort of automatically... 
My understanding is that doing the same on Google Cloud would require opening the terminal and enter some lines of command (something along this way, I suppose: https://github.com/timoschlueter/nightscout-librelink-up#variant-2-local). However - as I said already - I have no idea what would be the correct way to do it...

**Reply by Navid200** [reply]
Thanks for the link.
Let me see if we do anything about this.

**Reply by Navid200** [reply]
OK, now that we have Tzachi's feedback, this is what can offer you for now.

I am happy to walk you through the installation, step by step.
You don't have to do it.  I present the first step.  If you are interested to proceed, after you complete step one, I will present step two.

1- Post a screenshot of the status page so that I can see what version of our platform you are on: https://navid200.github.io/xDrip/docs/Nightscout/Status.html

**Reply by mduiella** [reply]
Dear @Navid200,
I'm really grateful for your support, much appreciated!
Please see below the screenshot of the status page.
best,
Matteo

![Screenshot 2023-09-05 164722](https://github.com/NightscoutFoundation/xDrip/assets/103216248/8c397248-083a-473f-aa78-f3cbe72f0d35)

**Reply by Navid200** [reply]
I was wrong.
Heroku is expansive and can make life easier.

I am sorry about my mistake.
Can you ask Nightscout pro if they have this feature?  You have Heroku for $5 a month as an option.  If I am not mistaken, Nightscout pro is $4 a month.
<br/>  
  
---  

With respect to doing this on Google Cloud, I cannot test this because I don't have Libre.  
So, what I provide here is just a suggestion for someone who is willing to test.  I cannot make any promises.  
  
If I was to do it, the first thing I would do would be to create a backup of the database: https://navid200.github.io/xDrip/docs/Nightscout/DatabaseBackup.html

Let's go over what this is and what it isn't.
If you create a backup of your Nightsocut this way and transfer the backup file to a computer so that it is not on the virtual machine, you will be able to restore the database in case something catastrophic happens to your setup.
However, the backup will only restore what you have up to the time you create the backup.
If you create the backup today and something goes wrong in a month from now, if you restore the backup, it will not restore the data you have accumulated over the past month.  That's why if you suspect something may go wrong, you should create frequent backups.

What I am going to explain for the remaining steps is only based on what I think is needed and is not base on experience.  I don't have Libre.  So, I cannot test this myself.
<br/>  
  
---  
  

After you have created a backup, you need to create a local pull from the script repository.
To do that, you need to exit the menu in Google Cloud Nightsocut, which is the last item on the main menu.

After existing the menu, you will be in a shell.  You need to type and enter this:
```cd /```  

Then, type and enter this to create a new directory:
```sudo mkdir LibreLinkup```  
  
Then, type and enter this to enter the new directory:
```cd LibreLinkup```
   
Then, enter this to create a local repo:
```sudo git clone https://github.com/timoschlueter/nightscout-librelink-up.git```

Now, enter this to change directory to the repository directory that was just imported:
```cd nightscout-librelink-up```

The next line may not be necessary:
```sudo git checkout main```  
  
Now, enter this to install the script:
```sudo npm install```  

<br/>  
  
---  
  
Now, copy the file from the link you provided to your computer:   
![Screenshot 2023-09-05 130718](https://github.com/NightscoutFoundation/xDrip/assets/51497406/70d7fe78-9700-472d-b192-dcee36ae68f7)
  
Edit the file on your computer and replace the labels with your details being user name, password and hostname.  
Save the file as start.sh.

Upload the file to your virtual machine.  This explains how to upload a backup: https://navid200.github.io/xDrip/docs/Nightscout/DatabaseRestore.html
But here, you are uploading the file named start.sh that you created.  
  
Enter:
```cd```

This takes you to your home directory where the file you uploaded will be.  So, in that directory you should see start.sh if you enter:
```ls```

Enter:
```sudo chmod 755 start.sh```
  
---  
  
To start the process, to ensure you are in home directory, enter
```cd```

Then, enter:
```./start.sh```

<br/>  

This is just for testing.  Every time, Google Cloud Nightscout restarts, you will need to start.sh again.

**Reply by mduiella** [reply]
Dear @Navid200,

No need to apologise! I wasn't very clear in my initial message, to be fair...
I am really grateful for your help, and we are all in debt for the amazing work that you and the other developers have been doing!

Yes, I have seen in some other discussion on fb groups that Nightscout pro has this feature (or rather, it's possible to have it added). 

Still, I am happy to act as tester for NS on Google Cloud, as I believe this can be useful for an increasing number of people (at least, until a way is found for XDrip to get readings directly from FS3).
Also, our Nightscout is not really essential (my wife doesn't really need to be monitored by others, nor does she use a pump), so if I mess up and everything crashes, it's no big deal. 

At first glance, your instructions seem very clear: I give it a try and let you know how it goes...

**Reply by mduiella** [reply]
Dear @Navid200,

all went smoothly, up to the last command:
after having entered: `start.sh` 
I got the following message: `-bash: start.sh: command not found`

**Reply by Navid200** [reply]
Did you upload the file?
If yes, can you see it if you type ls in your home directory?

**Reply by Navid200** [reply]
Wait, I made a mistake.  Just a sec.  Let me correct a mistake.

**Reply by Navid200** [reply]
I edited the line in the instructions.  Instead of entering start.sh, you need to enter ./start.sh

**Reply by mduiella** [reply]
thanks,
yes, I had uploaded the file (with my LibreLinkUp username and password and Nightscout url), and it appears correctly under the home directory.
Now, after entering `./start.sh` I get the following message:
`-bash: ./start.sh: /bin/bash^M: bad interpreter: No such file or directory`

**Reply by Navid200** [reply]
What do you see if you type this assuming nothing private is shown?

```ls -lt```

**Reply by mduiella** [reply]
I get the following:
![Screenshot 2023-09-05 222747](https://github.com/NightscoutFoundation/xDrip/assets/103216248/d5e8519a-07ac-42e4-a83f-517c758346d1)

(yes, I had to upload the file twice, as the first time it had kept the wrong .txt extension...)

**Reply by Navid200** [reply]
Enter:
```nano start.sh```

Take a screenshot.
Press ctrl X together to exit nano.

Edit the image.  Blackout user id, password and hostname.
Then, post the edited screenshot.

**Reply by mduiella** [reply]
here it is
![Screenshot 2023-09-05 223630](https://github.com/NightscoutFoundation/xDrip/assets/103216248/7dd45fef-8252-4b74-aa2a-680ece0c90ea)

**Reply by Navid200** [reply]
What if you try this?

```sudo ./start.sh```

**Reply by mduiella** [reply]
same error message: `sudo: unable to execute ./start.sh: No such file or directory`

**Reply by Navid200** [reply]
Let's create a token.
Open your Nightsocut in a browser.
On the following page, I have described how to create a readable toke: https://navid200.github.io/xDrip/docs/Nightscout/NS_Followers.html

Instead of following that to the letter, enter admin instead of readable.
You will end up with something like this:
![Screenshot 2023-09-05 171413](https://github.com/NightscoutFoundation/xDrip/assets/51497406/b916f45b-6ddc-4c14-bafe-dd16d33ca411)
I entered lib as name.  The rest was added by Nightscout.

Now that you have an admin token.  Copy the entire token.  In my example, it would be lib-dadd2510b593d3a6.
Just drag with your mouse to copy it.

Now, go back to the virtual machine and enter:
```sudo nano start.sh```
 
Replace foo-bar-baz with the token you copied.
press ctrl and O.  It will ask you to enter a name to write to.  But, it already has the file name as default.  Just press enter to save.  Now, press ctrl and X together to exit nano.

Try again running start.sh as we talked about before.  Does it make any difference?

**Reply by mduiella** [reply]
No, I still get the same error message

**Reply by Navid200** [reply]
Can I see a screenshot of the error message?

**Reply by mduiella** [reply]
here it is:
![image](https://github.com/NightscoutFoundation/xDrip/assets/103216248/af6c2125-507a-4896-96ae-cfb60fc6e5f3)

and this is how I have edited the file:
![image](https://github.com/NightscoutFoundation/xDrip/assets/103216248/b8fb6a5d-a0e6-48c3-9418-cf2252655ff8)

**Reply by Navid200** [reply]
This means the file start.sh is not in the directory you are in.

**Reply by mduiella** [reply]
yes, it makes sense... 
but (I think) I am in the home directory, and the file seems to be there:
![image](https://github.com/NightscoutFoundation/xDrip/assets/103216248/dd7fd451-2774-45da-a72d-9cf8b6270538)

**Reply by Navid200** [reply]
I'm sorry I don't know what is causing this.

**Reply by mduiella** [reply]
No problem :)
we gave it a try... 
I need to log off now, many thanks again for having taken the time!

**Reply by mduiella** [reply]
Dear @Navid,

FYI, I gave it another try, and now it sort-of worked.
First, I tried moving the start.sh file from the home directory to the `/LibreLinkup/nightscout-librelink-up/` directory, but I still couldn't manage to run it.
So, I decided to enter the commands manually one by one (that is, exporting the variables and then executing the command `npm start`).
This time, it seems to have worked, as the programme is logging in to LibreLinkUP to retrieve the data every 5 mins.

I would still have one question: above you said that this was just for testing, and I would have to start.sh again every time Google Cloud Nightscout restarts. So, what would one need to do to have this permanently set up?

Many thanks again for your support!
Matteo

**Reply by Navid200** [reply]
Thanks for getting it going?

I don't have an answer to your question.  What I stated was just speculation.  I have no way of testing this because I have no Libre.
The person who has written the script can tell you what you should do differently so that start.sh runs as intended.  And he can tell you if restarting will work or not.

**Reply by mduiella** [reply]
ok, I will inquire with the person who developed the script,
Many thanks again for your support!

**Reply by Sleepy46** [reply]
@mduiella  can you explain your solution exactly? 
I have the same Problem with the Uploader

**Comment by Navid200** [comment]
Nightscout is open source.  There may be more than one repository providing Nightsocut.
Nightscout has capabilities and documentation and those may differ from one repository to another.

In order to run Nightscout, you need a server.  There are several server options.

Google Cloud Nightscout is one server option.  Google Cloud Nightscout is not a version of Nightscout.

If you have questions about Google Cloud Nightscout, we can help.  If you have questions about Nightscout itself, this is not the best place for asking.  It doesn't hurt.  But, you may get more relevant answers asking in Nightsocut repositories or support forums.

**Comment by tzachi-dar** [comment]
There is an option for xDrip to work with libre3 without the site that you have mentioned.
See https://androidaps.readthedocs.io/lt/latest/Hardware/Libre3.html for more details.

Alternatively, deploying the app you want on a google server is probably possible, but might mess up your server.
So, first open another vm (might be on another project).
Than you need to deploy the app mentioned like every other js app on that server.
You might see how nightscout is installed, but if you are not a programmer, this might be hard for you to do.
I suggest that you will stick with the first option.

**Reply by mduiella** [reply]
Thanks,
yes, I had seen also this possibility with Juggluco, but it's only for Android.

**Comment by navid** [comment]
i am the wrong navid btw
